# <center> Merge() | Join() | Concat()</center>

<img src="https://files.realpython.com/media/Merging-Joining-and-Concatenating-in-Pandas_Watermarked_1.e45698a508a4.jpg" height=600 width=600/>

<center>
[Real Python](https://files.realpython.com)
</center>

## Joining DataFrames in Pandas

## Table of Content

1. [Introduction](#section1)
2. [Library import](#section2)
3. [Create DataFrame](#section3)
    - 3.1 [Concatenate DataFrames](#section301)
    - 3.2 [Merge DataFrames](#section302)
    
4. [With Real Life dataset](#section4)
    - 4.1 [User usage data](#section401)
    - 4.2 [User Device Data](#section402)
    - 4.3 [Different devices data](#403)
5. [Problem Statement](#section5)
6. [Different merge operations](#section6)
7. [Setps to Follow](#section7)
    - 7.1 [Inner Merge / Inner Join](#section701)
    - 7.2 [Left merge / Left join](#section702)
    - 7.3 [Right merge / Right join](#section703)
    - 7.4 [Outer merge / Full outer join](#section704)
    
8. [Comming back to our original problem of phones](#section8)

<a id="section1"></a>
## 1. Introduction


__Joining__ and __merging DataFrames__ is the core process to start with __data analysis__ and __machine learning__ tasks. It is one of the __toolkits__ which every __Data Analyst or Data Scientist__ should master because in almost all the cases data comes from __multiple source__ and __files__.

-  When you want to build some machine learning model on a data came from __multiple sources__ you need you may need to merge multiple csv files together in a __single DataFrame__.

- Thankfully you have the most popular library in python, __pandas__ to your rescue! pandas provides various facilities for easily combining together __Series__, __DataFrames__, and __Panel objects__ with various kinds of set logic for the indexes and relational algebra functionality in the case of __join / merge-type__ operations.

To join these DataFrames, pandas provides multiple functions like 
- <font color='blue'>concat()</font>
- <font color='blue'>merge()</font>
- <font color='blue'>join()</font>

<a id='section2'></a>
## 2. Import library

In [1]:
import pandas as pd
import numpy as np

<a id="section3"></a>
## 3. Create dataframe to perform different operations

In [2]:
dummy_data1 = {
        'id': ['1', '2', '3', '4', '5'],
        'Feature1': ['A', 'C', 'E', 'G', 'I'],
        'Feature2': ['B', 'D', 'F', 'H', 'J']}

df1 = pd.DataFrame(dummy_data1, columns = ['id', 'Feature1', 'Feature2'])

df1

id Feature1 Feature2
0  1        A        B
1  2        C        D
2  3        E        F
3  4        G        H
4  5        I        J

In [3]:
dummy_data2 = {
        'id': ['1', '2', '6', '7', '8'],
        'Feature1': ['K', 'M', 'O', 'Q', 'S'],
        'Feature2': ['L', 'N', 'P', 'R', 'T']}
df2 = pd.DataFrame(dummy_data2, columns = ['id', 'Feature1', 'Feature2'])

df2

id Feature1 Feature2
0  1        K        L
1  2        M        N
2  6        O        P
3  7        Q        R
4  8        S        T

In [4]:
dummy_data3 = {
        'id': ['1', '2', '3', '4', '5', '7', '8', '9', '10', '11'],
        'Feature3': [12, 13, 14, 15, 16, 17, 15, 12, 13, 23]}
df3 = pd.DataFrame(dummy_data3, columns = ['id', 'Feature3'])

df3

id  Feature3
0   1        12
1   2        13
2   3        14
3   4        15
4   5        16
5   7        17
6   8        15
7   9        12
8  10        13
9  11        23

<a id="section301"></a>
### 3.1 Concatenate DataFrames

In [5]:
df_row = pd.concat([df1, df2],ignore_index=True)

df_row

id Feature1 Feature2
0  1        A        B
1  2        C        D
2  3        E        F
3  4        G        H
4  5        I        J
5  1        K        L
6  2        M        N
7  6        O        P
8  7        Q        R
9  8        S        T

<a id="section302"></a>
### 3.2 Merge DataFrames


<center><img src="https://storage.googleapis.com/kagglesdsdata/datasets%2F784297%2F1347930%2Fdata%20Merge.png?GoogleAccessId=databundle-worker-v2@kaggle-161607.iam.gserviceaccount.com&Expires=1595219747&Signature=s1q7hSdnci%2BMvhHymlBQq5idowG4%2BXokagmIMOAWMlt86xD2iPbGT65z2v7x%2FS%2B1HKWaaqqrUlyFSCebIWFqJ4EyuVP5LcfZ01E9HzwW751DBq0glWpJZkAO7HTGR8OcNSbgLks%2FjgaM%2BTGhAYrG7ZiY4EeSYoi9G764XQyUCdMWGpP5%2FNJysES6XoJtmFivN9PlM8sa1NKCqgqw%2FQHi8xdw701q2pEqrkfpIzqD3iGybV1f9WdCYa0zjCAF%2BPSRWHD%2Fw9Tqwnw9YfFsZu%2FgNdHjkiAgupe9pGJatda7uZO0U8Hsqb8c0Sofbvg0U0tc2YkpJ8Fywzd%2BVn5QPkeDiw%3D%3D"/></center>

<img src="https://shanelynnwebsite-mid9n9g1q9y8tt.netdna-ssl.com/wp-content/uploads/2017/03/pandas-merge-join-different-variable-names-copy-e1488722312527.jpg"/>

DataFrames df_row (which you created by concatenating __df1__ and __df2__ along the row) and __df3__ on the common column (or key) __id__.

### Case 1

In [6]:
df_merge_col=pd.merge(df_row,df3,on='id')

In [7]:
df_merge_col

id Feature1 Feature2  Feature3
0  1        A        B        12
1  1        K        L        12
2  2        C        D        13
3  2        M        N        13
4  3        E        F        14
5  4        G        H        15
6  5        I        J        16
7  7        Q        R        17
8  8        S        T        15

You can notice that the DataFrames are now merged into a single DataFrame based on the __common values present__ in the __id column__ of both the DataFrames. 
    - For example, here id value 1 was present with both A, B and K, L in the DataFrame df_row hence this id got repeated twice in the final DataFrame df_merge_col with repeated value 12 of Feature3 which came from DataFrame df3.

[Read here more cases of concat](https://pandas.pydata.org/pandas-docs/stable/user_guide/merging.html)

### Case 2

It might happen that the column on which you want to merge the DataFrames have different names (unlike in this case). For such merges, you will have to specify the arguments __left_on__ as the left DataFrame name and __right_on__ as the right DataFrame name, like :

In [8]:
df_merge_difkey = pd.merge(df_row, df3, left_on='id', right_on='id')

df_merge_difkey

id Feature1 Feature2  Feature3
0  1        A        B        12
1  1        K        L        12
2  2        C        D        13
3  2        M        N        13
4  3        E        F        14
5  4        G        H        15
6  5        I        J        16
7  7        Q        R        17
8  8        S        T        15

<a id="section4"></a>
## 4. With Real Life dataset

Let's try to get the better and deeper intution using some some practical data set.

- To explain this we have taken some real data from the __KillBiller application__.

<a id="section401"></a>
### 4.1  User usage data

- Dataset containing users monthly mobile usage statistics

[See the complete blog Better to read from this blog](https://www.shanelynn.ie/merge-join-dataframes-python-pandas-index-1/)

In [9]:
user_usage=pd.read_csv('https://raw.githubusercontent.com/shanealynn/Pandas-Merge-Tutorial/master/user_usage.csv')
user_usage.head()

outgoing_mins_per_month  outgoing_sms_per_month  monthly_mb  use_id
0                    21.97                    4.82     1557.33   22787
1                  1710.08                  136.88     7267.55   22788
2                  1710.08                  136.88     7267.55   22789
3                    94.46                   35.17      519.12   22790
4                    71.59                   79.26     1557.33   22792

In [10]:
# user_usage shape
user_usage.shape

(240, 4)

<a id="section402"></a>
### 4.2 User Device Data

- Dataset containing details of an individual “use” of the system, with dates and device information.

In [11]:
# user device data
user_device=pd.read_csv('https://raw.githubusercontent.com/shanealynn/Pandas-Merge-Tutorial/master/user_device.csv')
user_device.head()

use_id  user_id platform  platform_version     device  use_type_id
0   22782    26980      ios              10.2  iPhone7,2            2
1   22783    29628  android               6.0    Nexus 5            3
2   22784    28473  android               5.1   SM-G903F            1
3   22785    15200      ios              10.2  iPhone7,2            3
4   22786    28239  android               6.0  ONE E1003            1

In [12]:
# user_device shape
user_device.shape

(272, 6)

<a id="section403"></a>
### 4.3 Different devices data

- Dataset with device and manufacturer data, which lists all __Android devices__ and their __model code__.

In [13]:
device=pd.read_csv('https://raw.githubusercontent.com/shanealynn/Pandas-Merge-Tutorial/master/android_devices.csv')
device.head()

Retail Branding Marketing Name    Device                      Model
0             NaN            NaN    AD681H  Smartfren Andromax AD681H
1             NaN            NaN     FJL21                      FJL21
2             NaN            NaN       T31              Panasonic T31
3             NaN            NaN  hws7721g         MediaPad 7 Youth 2
4              3Q        OC1020A   OC1020A                    OC1020A

In [14]:
# device shape
device.shape

(14546, 4)

### Important Note: 

There are __linking__ attributes between the sample datasets that are important to note –: __use_id__ is shared between the __user_usage__ and __user_device__, and the __device column__ of __user_device__ and __Model__ column of the __devices dataset__ contain common codes.

<a id="section5"></a>
## 5. Problem Statement 

This particular explaination is take form [shanelynn blog](https://www.shanelynn.ie/merge-join-dataframes-python-pandas-index-1/) . For more indepth knowledge check his blog.

<center><img src="https://qph.fs.quoracdn.net/main-qimg-968c26fb806e35f0f0477328db6f72d8" height=400 width=400/></center>

Using all above three dataset we would like to determine if the usage patterns for __users differ__ between __different devices__. For example, do users using <font color='red'>__Samsung devices__</font> use __more__ call minutes than those using <font color='red'>__LG devices__</font>? This is a toy problem given the small sample size in these dataset, but is a perfect example of __where merges are required__.

__Question__: 

We want to form a single dataframe with columns for user usage figures (__calls per month__, __sms per month__ etc) and also columns with device information (__model, manufacturer__, etc). We will need to __merge__ (or __join__) our sample datasets together into one single dataset for analysis.

## 6. Different merge operations 

<center><img src="https://storage.googleapis.com/kagglesdsdata/datasets%2F784297%2F1347930%2Fmerge_ops.jpg?GoogleAccessId=databundle-worker-v2@kaggle-161607.iam.gserviceaccount.com&Expires=1595224271&Signature=WPlHfSH3a2RdTi2TbBp3gGBf8VuMriyBFRTh5VaChWYLQfcV8ekmET%2Fhff%2Fq3%2FQmn9bRh%2BuaWtXa%2BfoY3edztz2QrPK2aX6Z3TtXHarId4D%2BHUeJEH8EuQbRJ1RAKhumxmDcjSM40UwxrSHzXDjHcXHyyzFR814jzUmNJ7uq8uiIsrBzBXQKoqVDMp%2FLRqrpXY%2F9y9sQiwBtJSvee825v4yexpiPpJZ%2FBLuxZQC%2BcDfnGTDoCP1p5mq8MdpdR4XgUE4GFe0yJS1eVn5%2FyIzkgqlf8TF0p17aOQMsg1cvLUfRkDVZR2yNqFtssxPfVzf1d43gxtSUJOxfn5lVzO5Cgw%3D%3D"/></center>

In [15]:
data1=pd.merge(user_usage,user_device[['use_id', 'platform', 'device']],on='use_id',how='outer',indicator=True)
data1.head()

outgoing_mins_per_month  outgoing_sms_per_month  monthly_mb  use_id  \
0                    21.97                    4.82     1557.33   22787   
1                  1710.08                  136.88     7267.55   22788   
2                  1710.08                  136.88     7267.55   22789   
3                    94.46                   35.17      519.12   22790   
4                    71.59                   79.26     1557.33   22792   

  platform    device _merge  
0  android  GT-I9505   both  
1  android  SM-G930F   both  
2  android  SM-G930F   both  
3  android     D2303   both  
4  android  SM-G361F   both

In [16]:
data1.shape

(353, 7)

In [17]:
data1.head()

outgoing_mins_per_month  outgoing_sms_per_month  monthly_mb  use_id  \
0                    21.97                    4.82     1557.33   22787   
1                  1710.08                  136.88     7267.55   22788   
2                  1710.08                  136.88     7267.55   22789   
3                    94.46                   35.17      519.12   22790   
4                    71.59                   79.26     1557.33   22792   

  platform    device _merge  
0  android  GT-I9505   both  
1  android  SM-G930F   both  
2  android  SM-G930F   both  
3  android     D2303   both  
4  android  SM-G361F   both

<a id="section6"></a>
## 7. Setps to Follow 

- In this example scenario, we will need to perform the following steps:

#### Step 1:

For each row in the __user_usage__ dataset – make a new column that contains the __device__ code from the __user_devices__ dataframe. i.e. for the first row, the __use_id__ is __22787__, so we go to the __user_devices__ dataset, find the __use_id 22787__, and copy the value from the __device__ column across.

#### Step 2: 

After this is complete, we take the new device columns, and we find the corresponding __Retail Branding__ and __Model__ from the devices dataset.

#### Step 3: 

Finally, we can look at different statistics for usage splitting and grouping data by the __device manufacturers__ used.

#### <font color='red'> Question : Can I use a for loop?</font>

__Yes__. You could write for loops for this task. The first would loop through the use_id in the user_usage dataset, and then find the right element in user_devices. The second for loop will repeat this process for the devices.

However, using for loops will be much slower and more verbose than using Pandas merge functionality. So,  if you come across this situation – __don’t use for loops__.

<a id="section701"></a>
### 7.1 Inner Merge / Inner Join

- Merging user_usage with user_devices

- Lets see how we can correctly add the __device__ and __platform__ columns to the __user_usage__ dataframe using the Pandas Merge command.

In [18]:
result = pd.merge(user_usage,
                 user_device[['use_id', 'platform', 'device']],
                 on='use_id')

result.head()

outgoing_mins_per_month  outgoing_sms_per_month  monthly_mb  use_id  \
0                    21.97                    4.82     1557.33   22787   
1                  1710.08                  136.88     7267.55   22788   
2                  1710.08                  136.88     7267.55   22789   
3                    94.46                   35.17      519.12   22790   
4                    71.59                   79.26     1557.33   22792   

  platform    device  
0  android  GT-I9505  
1  android  SM-G930F  
2  android  SM-G930F  
3  android     D2303  
4  android  SM-G361F

In [19]:
print(result.shape)

(159, 6)


<center><img src="https://shanelynnwebsite-mid9n9g1q9y8tt.netdna-ssl.com/wp-content/uploads/2017/02/pd-merge_intro_image-e1488724380571.png"/></center>

__Observation__ : __Decrement__ in the number of __rows__

- Result of merging __user_usage__ with user __user_device__ based on a common column.


The operation above we perform above is called <font color ='red'>__Inner merge__</font>
-  The merging operation at its simplest takes a left dataframe (the first argument), a right dataframe (the second argument), and then a merge column name, or a column to merge “on”. In the output/result, rows from the left and right dataframes are matched up where there are common values of the merge column specified by “on”.


<center><h3> Through Venn Diagram(Inner Merge)</h3></center>

<center><img src="https://images.squarespace-cdn.com/content/v1/5732253c8a65e244fd589e4c/1464122775537-YVL7LO1L7DU54X1MC2CI/ke17ZwdGBToddI8pDm48kMjn7pTzw5xRQ4HUMBCurC5Zw-zPPgdn4jUwVcJE1ZvWMv8jMPmozsPbkt2JQVr8L3VwxMIOEK7mu3DMnwqv-Nsp2ryTI0HqTOaaUohrI8PIvqemgO4J3VrkuBnQHKRCXIkZ0MkTG3f7luW22zTUABU/image-asset.png" height=250 width=250/></center>

### Note: 
The merging operation at its simplest takes a __left dataframe__ (the first argument), a __right dataframe__ (the second argument), and then a __merge column name__, or a column to merge __on__. 
- In the __output/result__, rows from the left and right dataframes are matched up where there are common values of the merge column specified by “on”.

<a id="section702"></a>
### 7.2  Left merge / Left join

Left join example in pandas. Specify the join type in the __how__ command. A __left join__, or __left merge__, keeps every row from the left dataframe.

<center><img src="https://images.squarespace-cdn.com/content/v1/5732253c8a65e244fd589e4c/1464122797709-C2CDMVSK7P4V0FNNX60B/ke17ZwdGBToddI8pDm48kMjn7pTzw5xRQ4HUMBCurC5Zw-zPPgdn4jUwVcJE1ZvWEV3Z0iVQKU6nVSfbxuXl2c1HrCktJw7NiLqI-m1RSK4p2ryTI0HqTOaaUohrI8PIO5TUUNB3eG_Kh3ocGD53-KZS67ndDu8zKC7HnauYqqk/image-asset.png" height=300 width=300/>

- For our __left merge__, we expect the result to have the same number of rows as our __left dataframe__ “user_usage” (240), with missing values for all but 159 of the merged “platform” and “device” columns (81 rows).


- In __merge operations__ where a single row in the __left dataframe__ is matched by __multiple rows__ in the __right dataframe__, multiple result rows will be generated. 

    i.e. if a __use_id__ value in __user_usage__ appears __twice__ in the __user_device__ dataframe, there will be __two rows__ for that __use_id__ in the join result.

In [20]:
result = pd.merge(user_usage,
                 user_device[['use_id', 'platform', 'device']],
                 on='use_id', 
                 how='left')

In [21]:
result.tail()

outgoing_mins_per_month  outgoing_sms_per_month  monthly_mb  use_id  \
235                   260.66                   68.44      896.96   25008   
236                    97.12                   36.50     2815.00   25040   
237                   355.93                   12.37     6828.09   25046   
238                   632.06                  120.46     1453.16   25058   
239                   488.70                  906.92     3089.85   25220   

    platform device  
235      NaN    NaN  
236      NaN    NaN  
237      NaN    NaN  
238      NaN    NaN  
239      NaN    NaN

- Result from left-join or left-merge of two dataframes in Pandas. Rows in the left dataframe that have __no corresponding join value__ in the right dataframe are left with __NaN__ values.

<a id="section703"></a>
### 7.3 Right merge / Right join

We can repeat this process with a __right join / right merge__, simply by replacing __how=left__ with __how=right__ in the Pandas merge command.



<center><img src="https://images.squarespace-cdn.com/content/v1/5732253c8a65e244fd589e4c/1464122744888-MVIUN2P80PG0YE6H12WY/ke17ZwdGBToddI8pDm48kMjn7pTzw5xRQ4HUMBCurC5Zw-zPPgdn4jUwVcJE1ZvWlExFaJyQKE1IyFzXDMUmzc1HrCktJw7NiLqI-m1RSK4p2ryTI0HqTOaaUohrI8PI-FpwTc-ucFcXUDX7aq6Z4KQhQTkyXNMGg1Q_B1dqyTU/image-asset.png" height=300 width=300/>

In [22]:
result = pd.merge(user_usage,
                 user_device[['use_id', 'platform', 'device']],
                 on='use_id', 
                 how='right')

In [23]:
result.head()

outgoing_mins_per_month  outgoing_sms_per_month  monthly_mb  use_id  \
0                    21.97                    4.82     1557.33   22787   
1                  1710.08                  136.88     7267.55   22788   
2                  1710.08                  136.88     7267.55   22789   
3                    94.46                   35.17      519.12   22790   
4                    71.59                   79.26     1557.33   22792   

  platform    device  
0  android  GT-I9505  
1  android  SM-G930F  
2  android  SM-G930F  
3  android     D2303  
4  android  SM-G361F

In [24]:
result.tail()

outgoing_mins_per_month  outgoing_sms_per_month  monthly_mb  use_id  \
267                      NaN                     NaN         NaN   23047   
268                      NaN                     NaN         NaN   23048   
269                      NaN                     NaN         NaN   23050   
270                      NaN                     NaN         NaN   23051   
271                      NaN                     NaN         NaN   23052   

    platform         device  
267      ios      iPhone7,1  
268  android  ONEPLUS A3003  
269      ios      iPhone7,2  
270      ios      iPhone7,2  
271      ios      iPhone8,4

The result expected will have the same number of rows as the __right dataframe__, __user_device__, but have several empty, or __NaN__ values in the columns originating in the __left dataframe__, __user_usage__ (namely “outgoing_mins_per_month”, “outgoing_sms_per_month”, and “monthly_mb”).
- Conversely, we expect no missing values in the columns originating in the right dataframe, “user_device”.

<a id="section704"></a>
### 7.4 Outer merge / Full outer join

<center><img src="https://images.squarespace-cdn.com/content/v1/5732253c8a65e244fd589e4c/1464122981217-RIYH5VL2MF1XWTU2DKVQ/ke17ZwdGBToddI8pDm48kMjn7pTzw5xRQ4HUMBCurC5Zw-zPPgdn4jUwVcJE1ZvWEV3Z0iVQKU6nVSfbxuXl2c1HrCktJw7NiLqI-m1RSK4p2ryTI0HqTOaaUohrI8PIO5TUUNB3eG_Kh3ocGD53-KZS67ndDu8zKC7HnauYqqk/image-asset.png" height=300 width=300/>

An __outer merge__ using __Pandas__, also referred to as a __full outer join__ or just __outer join__. An outer join can be seen as a combination of left and right joins, or the opposite of an inner join. In outer joins, every row from the left and right dataframes is retained in the result, with NaNs where there are no matched join variables.

- Outer merge result using Pandas. Every row from the __left__ and __right__ dataframes is retained in the result, with missing values or numpy NaN values where the merge column doesn’t match.

In [25]:
result = pd.merge(user_usage,
                 user_device[['use_id', 'platform', 'device']],
                 on='use_id', 
                 how='outer', 
                 indicator=True)

### Using merge indicator to track merges

To assist with the identification of where rows originate from, Pandas provides an [indicator parameter](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.merge.html) that can be used with the merge function which creates an additional column called ___merge__ in the output that labels the original source for each row.

<a id="section8"></a>
### 8. Comming back to our original problem of phones 

In [26]:
print(user_usage.shape)
print(user_device.shape)

(240, 4)
(272, 6)


In [27]:
# First, add the platform and device to the user usage - use a left join this time.

result = pd.merge(user_usage,
                 user_device[['use_id', 'platform', 'device']],
                 on='use_id',
                 how='left')


In [28]:
result.head()

outgoing_mins_per_month  outgoing_sms_per_month  monthly_mb  use_id  \
0                    21.97                    4.82     1557.33   22787   
1                  1710.08                  136.88     7267.55   22788   
2                  1710.08                  136.88     7267.55   22789   
3                    94.46                   35.17      519.12   22790   
4                    71.59                   79.26     1557.33   22792   

  platform    device  
0  android  GT-I9505  
1  android  SM-G930F  
2  android  SM-G930F  
3  android     D2303  
4  android  SM-G361F

In [29]:
result.shape

(240, 6)

- At this point, the platform and device columns are included


- In the result along with all columns from user_usage

### Using left_on and right_on to merge with different column names

- The columns used in a merge operator do not need to be named the same in both the left and right dataframe. In the second merge above, note that the device ID is called “device” in the left dataframe, and called “Model” in the right dataframe.


- Different column names are specified for merges in Pandas using the __left_on__ and __right_on__ parameters, instead of using only the “on” parameter.


- Merging dataframes with different names for the joining variable is achieved using the left_on and right_on arguments to the pandas merge function.

<center><img src="https://shanelynnwebsite-mid9n9g1q9y8tt.netdna-ssl.com/wp-content/uploads/2017/03/pandas-merge-join-different-variable-names-copy-e1488722312527.jpg"/>

In [30]:
# Now, based on the "device" column in result, match the "Model" column in devices.
device.rename(columns={"Retail Branding": "manufacturer"}, inplace=True)
result = pd.merge(result, 
                  device[['manufacturer', 'Model']],
                  left_on='device',
                  right_on='Model',
                  how='left')



In [31]:
result.head()

outgoing_mins_per_month  outgoing_sms_per_month  monthly_mb  use_id  \
0                    21.97                    4.82     1557.33   22787   
1                  1710.08                  136.88     7267.55   22788   
2                  1710.08                  136.88     7267.55   22789   
3                    94.46                   35.17      519.12   22790   
4                    71.59                   79.26     1557.33   22792   

  platform    device manufacturer     Model  
0  android  GT-I9505      Samsung  GT-I9505  
1  android  SM-G930F      Samsung  SM-G930F  
2  android  SM-G930F      Samsung  SM-G930F  
3  android     D2303         Sony     D2303  
4  android  SM-G361F      Samsung  SM-G361F

Refrence: 

- https://www.shanelynn.ie/merge-join-dataframes-python-pandas-index-1/


Question : Where can i implement this knowledge?

- Check this competion/data and try to apply what you learn. 
    - [Competition to practice](kaggle.com/vin1234/starter-janatahack-healthcare-analytics)

If you guys have any suggestion for this notebook so we can build this notebook much better further let me know in the comment section.

<center>................**Waiting for your response**...................</center>
<br>

<center><img src='https://clientinsight.ca/wp-content/uploads/2018/03/waiting-for-phone.jpg' height=400 width=500/></center>